Fix bug in check required output size in _as_strided_scatter_meta

lantiankaikai opened the pull request:
Original Issue from #92670

pytest ./generated/test_XuyangBai_PointDSC.py -k test_004

==> RuntimeError: as_strided_scatter: sizes [4], strides [85], storage offset 256 and itemsize 4 requiring a storage size of 2048 are out of bounds for storage of size 1024

Repro:

```
class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        x[1].fill_diagonal_(0)   # this check size failed

device = torch.device("cpu")
model = Model()
model.to(device)

torch._dynamo.reset()
compiled_model = torch._dynamo.optimize("inductor")(model)

arg = [torch.rand([4, 1, 1])]
compiled_model(*arg)

```
The error was raised at the checking required size in as_strided_scatter. 

https://github.com/pytorch/pytorch/blob/master/torch/_prims/__init__.py#L1818

In the case of input is a tensor with storage offset(a view), when compute input's storage length, should also take input's base tensor's size/stride/offset into account instead of compare it with number of element of input.

This diff fix the bug and add test.

Fixes #ISSUE_NUMBER


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

lantiankaikai committed:
Fix bug in check required output size in _as_strided_scatter_meta

as_strided_scatter returns the clones of the input with fresh storage.

In case of input tensor with storage offset, when checking for size requirement, need to take input's origin tensor's size/stride/offset into account instead just compare it with number of element of input.

This diff fix the error and add test.

peterbell10 commented on torch/_prims/__init__.py:
```diff
@@ -1815,14 +1815,15 @@ def _as_strided_scatter_meta(
+    output_length = utils.compute_required_storage_length(size, stride, storage_offset)
     utils.check(
-        input.numel() >= required_size,
+        input_length >= output_length,
```
Do you have a repro using functionalization?

lantiankaikai replied:
updated with repro

lantiankaikai committed:
Update test_torchinductor_codegen_dynamic_shapes.py

xfails for test_as_strided_scatter_on_input_tensor_with_storage_offset_dynamic_shapes

lantiankaikai committed:
lint

peterbell10 commented on torch/_prims/__init__.py:
```diff
@@ -1814,14 +1814,17 @@ def _as_strided_scatter_meta(
+    input_length = utils.compute_required_storage_length(
+        input.size(), input.stride(), input.storage_offset()
+    )
+    output_length = utils.compute_required_storage_length(size, stride, storage_offset)
```
NIT: This quantity doesn't really have anything to do with the output tensor's geometry, so something like `required_length` or `view_length` would be a better fit.

lantiankaikai replied:
Thanks!  that make sense, i'll update the name

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -4505,6 +4505,23 @@ def fn(a, b):
+
+        inp = torch.randn(4, 1)
+        # as_strided_scatter in inductor is only making a clone of input
+        # and doesn't copy from src, clone src to itself for testing
```
It might be worth opening a separate issue. The copy is definitely in the lowering, so it would suggest inductor isn't treating the mutation properly.

lantiankaikai replied:
@yanboliang mentioned it might be intended to save operation as _as_strided_scatter_meta only need the output's tensor size info

yanboliang replied:
Actually what we are trying to fix is a prims bug rather than inductor bug, you can put the test in prims unit test folder. But since you found the inductor unit test bug, so I'm ok if you put it here. But your comment is not correct, you described what prims is doing, not what inductor is doing.

lantiankaikai commented:
The CI test failure in CUDA mode is due to in check_model_cuda, the input tensor's storage_offset is not preserved. 

https://github.com/pytorch/pytorch/blob/master/test/inductor/test_torchinductor.py#L431

So if input is a tensor view, the storage_offset is lost and thus test failed. 

In contrast, in CPU test the input tensor's size/stride/storage_offset are preserved. 

https://github.com/pytorch/pytorch/blob/master/test/inductor/test_torchinductor.py#L280

Is it a bug?

lantiankaikai committed:
preserve storage_offset for cuda test in test_torchinductor.py

lantiankaikai committed:
preserve storage_offset for cuda test in test_torchinductor.py

lantiankaikai committed:
Update comment in test_as_strided_scatter_on_input_tensor_with_storage_offset

lantiankaikai committed:
add test_as_strided_scatter to test_prims.py

lantiankaikai committed:
backup changes for test_inductor.py

lantiankaikai committed:
Merge branch 'master' into Inductor_test

ngimel commented on test/test_prims.py:
```diff
@@ -1094,6 +1094,35 @@ def test_reshape_view_method(self, device, dtype):
+
+        inp = make_arg((4, 1))
+        # as_strided_scatter in prim is only making a clone of input
+        # and doesn't copy from src, clone src of itself for testing
```
wdym? as_strided_scatter in prim is correctly coying the src:
```
In [26]: a=torch.ones(4,4)

In [27]: src = torch.ones(2,3)*2

In [28]: torch._prims.as_strided_scatter(a, src, (2,3), (3,1), 0)
Out[28]: 
tensor([[2., 2., 2., 2.],
        [2., 2., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])

lantiankaikai replied:
You are right, I think I miss interpret the _as_strided_scatter_meta as inside it doesn't copy the src.

https://github.com/pytorch/pytorch/blob/14ccad73b48edde1d4b41d26668c324424a531d7/torch/_prims/__init__.py#L1832

I'll update the comment.

ngimel commented on torch/_prims/__init__.py:
```diff
@@ -1814,14 +1814,19 @@ def _as_strided_scatter_meta(
+    )
     utils.check(
-        input.numel() >= required_size,
+        input_length >= required_view_length,
```
this is not correct, e.g. if I do `torch.as_strided_scatter(inp[1], (1, ), (1,), 2)` this is valid, but your check will fail.

lantiankaikai replied:
You raised a good point, the original code also failed with above test case. 
The behavior I observed in as_strided_scatter, when input is a view like x[1], it makes a copy of view's original tensor x, let's say x', applied src, then return the x'[1]. 

For example:

```
>>> x = torch.ones(4, 1)      // x = [[1],[1],[1],[1]]
>>> src = torch.ones(1)*2
>>> torch.as_strided_scatter(x[1], src, (1,), (1,), 0)  // x' =[[2],[1],[1],[1]], return x'[1] = [1]
tensor([1.])
>>> torch.as_strided_scatter(x[1], src, (1,), (1,), 1)  // x' =[[1],[2],[1],[1]], return x'[1] = [2]
tensor([2.])
>>> torch.as_strided_scatter(x[1], src, (1,), (1,), 2)  // x' =[[1],[1],[2],[1]], return x'[1] = [1]
tensor([1.])
```
If that's the expected behavior, I'm wondering how to check the size properly, as x[1] doesn't know the size of original x. Any suggestion?

lantiankaikai committed:
update size check _as_strided_scatter_meta to use untyped_storage

ngimel commented on torch/_prims/__init__.py:
```diff
@@ -1814,19 +1814,17 @@ def _as_strided_scatter_meta(
-        input.size(), input.stride(), input.storage_offset()
-    )
-    required_view_length = utils.compute_required_storage_length(
+    input_size = input.untyped_storage().size()
```
how would this work for fake tensors?

lantiankaikai replied:
I tested with original repro and observed that fake tensors can return the corrected size with input.untyped_storage().size(). 

I also test below:

```
>>> with FakeTensorMode():
...        x = torch.rand((4, 1))
...
>>> x
FakeTensor(FakeTensor(..., device='meta', size=(4, 1)), cpu)
>>> x.untyped_storage()
...
[torch.storage.UntypedStorage(device=meta) of size 16]

>>> x.untyped_storage().size()
16
>>> x[1].untyped_storage().size()
16
```

I can add a unit test in fake tensor. Do you see any concern using this to determine the base tensor's size of a view? Any suggestion on proper way to check the size? Appreciate your help!

ngimel commented on test/test_prims.py:
```diff
@@ -1094,6 +1094,46 @@ def test_reshape_view_method(self, device, dtype):
         self.assertEqual(result_eager, result_refs)
 
+    @dtypes(torch.float32)
+    def test_as_strided_scatter(self, device, dtype):
```
instead of adding one-off tests, it would make sense to add a few reference inputs to `as_strided_scatter` reference inputs in OpInfos in `common_methods_invocation.py`, then e.g. this error would have been discovered earlier.

lantiankaikai replied:
Thanks, let me look into that

lantiankaikai committed:
lint and unit test for fake tensor's untyped_storage size

lantiankaikai committed:
move as_strided_scatter test to common_methods_invocations.py

lantiankaikai committed:
Merge branch 'master' into Inductor_test

yanboliang commented:
@pytorchbot merge

lantiankaikai commented:
@pytorchbot label "topic: not user facing"

yanboliang commented:
@pytorchbot merge

lantiankaikai commented:
@pytorchbot merge

ngimel commented:
@pytorchbot merge -f "regular flow not working"

malfet commented:
@pytorchbot revert -m "Broke inductor, see https://hud.pytorch.org/hud/pytorch/pytorch/main/1?per_page=50&name_filter=inductor%2C%201%2C%201" -c ignoredsignal
